lib/models/package-info-cache.js

```diff
@@ -0,0 +1,1049 @@
+'use strict';
+
+/*
+ * Performance cache for information about packages (projects/addons/"apps"/modules)
+ * under an initial root directory and resolving addon/dependency links to other packages.
+ */
+
+const fs = require('fs-extra');
+const path = require('path');
+
+let realFilePathCache;
+let realDirectoryPathCache;
+
+/**
+ * Resolve the real path for a file, return null if does not
+ * exist or is not a file or FIFO, return the real path otherwise.
+ *
+ * @private
+ * @method getRealFilePath
+ * @param  {String} filePath the path to the file of interest
+ * @return {String} real path or null
+ */
+function getRealFilePath(filePath) {
+  let realPath;
+
+  try {
+    realPath = realFilePathCache[filePath];
+
+    if (realPath) {
+      return realPath;
+    }
+
+    let stat = fs.statSync(filePath);
+
+    if (stat.isFile() || stat.isFIFO()) {
+      realPath = fs.realpathSync(filePath);
+    }
+  } catch (e) {
+    if (
+      e !== null &&
+      typeof e === 'object' &&
+      (e.code === 'ENOENT' || e.code === 'ENOTDIR')
+    ) {
+      realPath = null;
+    } else {
+      throw e;
+    }
+  }
+
+  realFilePathCache[filePath] = realPath;
+  return realPath;
+}
+
+/**
+ * Resolve the real path for a directory, return null if does not
+ * exist or is not a directory, return the real path otherwise.
+ *
+ * @private
+ * @method getRealDirectoryPath
+ * @param  {String} directoryPath the path to the directory of interest
+ * @return {String} real path or null
+ */
+function getRealDirectoryPath(directoryPath) {
+  let realPath;
+
+  try {
+    realPath = realDirectoryPathCache[directoryPath];
+
+    if (realPath) {
+      return realPath;
+    }
+
+    let stat = fs.statSync(directoryPath);
+
+    if (stat.isDirectory()) {
+      realPath = fs.realpathSync(directoryPath);
+    }
+  } catch (e) {
+    if (
+      e !== null &&
+      typeof e === 'object' &&
+      (e.code === 'ENOENT' || e.code === 'ENOTDIR')
+    ) {
+      realPath = null;
+    } else {
+      throw e;
+    }
+  }
+
+  realDirectoryPathCache[directoryPath] = realPath;
+  return realPath;
+}
+
+/*
+ * Small utility class to store a single error during processing.
+ *
+ * @private
+ * @class ErrorEntry
+ */
+class ErrorEntry {
+  constructor(type, data) {
+    this.type = type;
+    this.data = data;
+  }
+}
+
+/*
+ * Small utility class to store a list of errors during processing.
+ * Instances of this exist in the PackageInfo and PackageInfoCache
+ * objects.
+ *
+ * @private
+ * @class ErrorEntry
+ */
+class ErrorList {
```

Comment:
needs unit test

Reply:
Several of the tests for PIC itself contain references to error lists that will fail if they aren't right. I guess I could add more specifically for ErrorList itself, but not sure if that's just overkill.